---
title: "IHM association with RA Disease Activity"
output:
  html_document:
    df_print: paged
    code_folding: hide
    toc: True
  html_notebook:
    toc: yes
editor_options:
  chunk_output_type: console
---

Using the RA multiomics dataset from https://www.nature.com/articles/s41597-022-01264-y

Figures 4f-g, Extended Data Figure 7c
```{r setup,message=F}
library(tidyverse)
library(ggpubr)
library(lme4)
source(file.path("scripts","utils.R"))

# IHM transcriptomic
ihm.sig <- readRDS(file.path("data","IHM","healthy_index_transcriptomic_signature_2021_04_19.rds"))

# IHM proteomics 
ihm.soma.sig <- read.csv(file.path("data","IHM","IHM.protein.signature.txt"),sep = "\t",strip.white = T,stringsAsFactors = F)

# RA-MAP microarray
ramap.rna.eSet <- readRDS(file.path("data","RA_MAP","microarray_eSet.rds"))
table(ramap.rna.eSet$CELL.TYPE)
ra.expr <- assayData(ramap.rna.eSet)$exprs
ra.clin <- phenoData(ramap.rna.eSet)@data
ra.clin$Tacera_ID <- as.character(ra.clin$Tacera_ID)
# vax HCs
vax.rna.eSet <- readRDS(file.path("data","RA_MAP","vaccine_microarray_eSet.rds"))
table(vax.rna.eSet$CELL.TYPE)
vax.expr <- assayData(vax.rna.eSet)$exprs
vax.clin <- phenoData(vax.rna.eSet)@data
vax.clin$TIME <- ifelse(vax.clin$TIME == "Visit_1","Baseline",vax.clin$TIME)

# merge
target.celltype <- "WB"
clin <- bind_rows(ra.clin,vax.clin)
table(clin$CELL.TYPE)
clin <- subset(clin,CELL.TYPE == target.celltype)
gplots::venn(list(ra=rownames(ra.expr),vax=rownames(vax.expr)))
common.genes <- intersect(rownames(ra.expr),rownames(vax.expr))
common.genes <- common.genes[common.genes != ""]
expr <- cbind(ra.expr[common.genes,],vax.expr[common.genes,])
expr <- expr[,rownames(clin)]

# RA-MAP somalogic
ramap.soma.eSet <- readRDS(file.path("data","RA_MAP","SOMASCAN_RA-Map_eSet.rds"))
ramap.soma.eSet <- ramap.soma.eSet[featureData(ramap.soma.eSet)@data[,"ColCheck"] == "PASS",ramap.soma.eSet$RowCheck == "PASS"]
protein.levels <- assayData(ramap.soma.eSet)$exprs
rownames(protein.levels) <- featureData(ramap.soma.eSet)$SomaId
protein.clin <- read.csv(file.path("data","RA_MAP","SOMASCAN_RA-Map_figshare_17_11_20.csv"),strip.white = T)
rownames(protein.clin) <- protein.clin$SampleId
protein.clin <- protein.clin[phenoData(ramap.soma.eSet)@data$SampleId,]
protein.clin$TimePoint <- gsub("6","6-",protein.clin$TimePoint)
```


### Calculate IHM surrogate scores
```{r score,out.width="900px"}
IHM_gene_score <- calculate_signature_score(expr,ihm.sig$positive,ihm.sig$negative)
microarray.samples <- cbind(clin,IHM_WB_score=IHM_gene_score[rownames(clin)])

IHM_protein_score <- calculate_signature_score(protein.levels,
                                            make.names(subset(ihm.soma.sig,direction == "positive")$SomaId),
                                            make.names(subset(ihm.soma.sig,direction == "negative")$SomaId))
somalogic.samples <- cbind(protein.clin,IHM_protein_score)

# only somalogic samples have response info 
samples <- merge(microarray.samples,somalogic.samples[,c("Patient_ID","TimePoint","SampleGroup","IHM_protein_score")],
                 by.x=c("Patient_ID","TIME"),by.y=c("Patient_ID","TimePoint"),all.x=T)
```

### Response to treatment
Only see grouping of responder and non-responder for csDMARD (UK-NHS standard of care first-line treatment)
```{r responding,out.width="900px"}
# somalogic
ggplot(protein.clin,aes(paste0(SampleGroup,"\n",TimePoint),IHM_protein_score)) +
  geom_boxplot(outlier.shape = NA) + geom_jitter(width=0.25,height = 0,alpha=0.5,aes(color=SampleGroup)) + scale_x_discrete(limits=rev) +
  theme_bw() + stat_compare_means(comparisons = list(c(1,2),c(3,4),c(1,3),c(2,4),c(4,5),c(3,5),c(2,5),c(1,5)),size=2) + ylim(NA,3) +
  xlab("Sample") + ylab("IHM protein surrogate score")

# use only patients with both timepoints
tmp <- names(which(table(subset(somalogic.samples,!is.na(SampleGroup) & !is.na(IHM_protein_score))$Patient_ID) == 2))
ggplot(subset(somalogic.samples,Patient_ID %in% tmp),aes(paste0(SampleGroup,"\n",TimePoint),IHM_protein_score)) +
  geom_hline(yintercept = median(subset(protein.clin,SampleGroup == "HC")$IHM_protein_score),linetype="dashed",alpha=0.5) +
  geom_boxplot(outlier.shape = NA) + scale_x_discrete(limits=rev) + 
  geom_line(aes(group=Patient_ID),alpha=0.3) +
  geom_point(size=1.5,alpha=0.5,aes(color=SampleGroup)) + ylim(NA,1.75) +
  theme_bw() + stat_compare_means(comparisons = list(c(1,2),c(3,4),c(1,3),c(2,4)),size=3,paired=F) + ggsci::scale_color_aaas() +
  xlab("Sample") + ylab("IHM protein surrogate score")
#ggsave("IHM.v.response.soma.pdf",height=3,width=4.25)

lm.res <- lmer(IHM_protein_score ~ AGE + GENDER + TimePoint:SampleGroup + (1|Patient_ID),subset(somalogic.samples,Patient_ID %in% tmp))
em <- emmeans::emmeans(lm.res,~ TimePoint:SampleGroup)
broom::tidy(emmeans::contrast(em,method="pairwise",adjust = "none"))

# gene expression
ggplot(subset(samples,!is.na(SampleGroup)),aes(paste0(SampleGroup,"\n",TIME),IHM_WB_score)) +
  geom_boxplot(outlier.shape = NA) + geom_jitter(width=0.25,height = 0,alpha=0.5,aes(color=SampleGroup)) + scale_x_discrete(limits=rev) +
  theme_bw() + stat_compare_means(comparisons = list(c(1,2),c(3,4),c(1,3),c(2,4),c(4,5),c(3,5),c(2,5),c(1,5)),size=2) + ylim(NA,3.5) +
  xlab("Sample") + ylab("IHM gene surrogate score")
  
# use only patients with both timepoints
tmp <- names(which(table(subset(samples,!is.na(SampleGroup) & !is.na(IHM_WB_score))$Patient_ID) == 2))
ggplot(as.data.frame(subset(samples,Patient_ID %in% tmp) %>% arrange(Patient_ID)),aes(paste0(SampleGroup,"\n",TIME),IHM_WB_score)) +
  geom_hline(yintercept = median(subset(samples,SampleGroup == "HC")$IHM_WB_score),linetype="dashed",alpha=0.5) +
  geom_boxplot(outlier.shape = NA) + scale_x_discrete(limits=rev) +  
  geom_line(aes(group=Patient_ID),alpha=0.3) +
  geom_point(size=1.5,alpha=0.5,aes(color=SampleGroup)) +
  theme_bw() + stat_compare_means(comparisons = list(c(1,2),c(3,4),c(1,3),c(2,4)),size=2.5,paired=F) + ggsci::scale_color_aaas() +
  xlab("Sample") + ylab("IHM transcriptional surrogate score")
#ggsave("IHM.v.response.gene.pdf",height=3,width=4.25)

lm.res <- lmer(IHM_WB_score ~ AGE + SEX + TIME:SampleGroup + (1|Patient_ID),subset(samples,Patient_ID %in% tmp))
em <- emmeans::emmeans(lm.res,~ TIME:SampleGroup)
broom::tidy(emmeans::contrast(em,method="pairwise",adjust = "none"))
```

### Association with disease activities
```{r da,out.width="900px"}
samples.baseline.melted <- reshape2::melt(subset(samples,TIME == "Baseline"),measure.vars = grep("score",colnames(samples)))

print(ggplot(samples.baseline.melted,aes_string("value","DAS28.0M")) + geom_point(alpha=0.75) + 
            geom_smooth(method="lm") + xlab("Baseline IHM score") + 
              stat_cor(method="spearman",cor.coef.name = "rho",size=2) + theme_bw() + facet_grid(~variable,scales = "free") + theme(aspect.ratio = 1))

```

### Session Info
```{r session}
sessionInfo()
```